src/reader.h

```diff
@@ -0,0 +1,29 @@
+/*
+ * Copyright 2017 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ **/
+#ifndef READER_H_
+#define READER_H_
+
+namespace google {
+
+class MetadataReader {
+ public:
+  virtual bool ValidateConfiguration() const;
```

Comment:
There's no reason to have a superclass for all readers, actually. I would keep this independently in `KubernetesReader` and `DockerReader`. Let's just revert https://github.com/Stackdriver/metadata-agent/pull/63/commits/89d6cf8cefd5e7095d8f64fd2512945585aa432e.

Reply:
The reader is the thing that's actually interacting with the endpoints. I'll talk with you offline about this to build additional context together with you, but when I tried implementing the validation in the updater it just resulted in transferred responsibility down to the reader.

Reply:
It's ok for each updater to pass the responsibility down to individual readers — that's why I suggested having a `ValidateConfiguration` function in each reader that needs it. I'm simply saying that it's not necessary to have the validation in the reader, and we should abstract that away. However, the polling updater needs to perform the validation before starting the polling threads, so let's document that in a virtual function.
Happy to discuss offline.

Reply:
Discussed and addressed offline.